toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -185,13 +313,206 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
 	if stopOnFailure && !cachingSucceeded {
 		return fmt.Errorf("failed to cache unresolved nodes")
 	}
+	return
+}
+
+// downloadAllAvailableDeltaRPMs scans a graph and for each build node in the graph and tries to replace it with a cached node instead.
+// to satisfy it.
+//   - realDependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - dependencyGraphDeltaCopy: A copy of the graph we will use to try to optimize the build nodes. This graph should be
+//     optimized to only contain the nodes we need to build.
+//   - cloner: The cloner to use to download the RPMs
+//   - stopOnFailure: If true, will stop the build if we fail to download any delta RPMs.
+func downloadAllAvailableDeltaRPMs(realDependencyGraph, dependencyGraphDeltaCopy *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner, stopOnFailure bool) (err error) {
+	timestamp.StartEvent("downloading delta nodes", nil)
+	defer timestamp.StopEvent(nil)
+
+	// First scan the copy of the graph we tried to optimize for all the SRPMs we need to build. We will use this list to
+	// match against all the nodes in the full graph.
+	srpmPaths := make(map[string]bool)
+	for _, n := range dependencyGraphDeltaCopy.AllBuildNodes() {
+		srpmPaths[n.SrpmPath] = true
+	}
+
+	// We don't want to download implicit nodes since they will be included in another node with the same SRPM. Keep a list
+	// of them so we can fix them up later.
+	skippedNodes := []*pkggraph.PkgNode{}
+	// We will need to keep track of the original path to delta path mapping so we can fix up the implicit nodes later.
+	originalPathToDeltaPathMap := make(map[string]string)
+
+	// For each build node, try to update it to a delta node with a downloaded RPM backing it.
+	logger.Log.Debugf("Resolving build nodes")
+	for _, n := range realDependencyGraph.AllBuildNodes() {
+		// Implicit nodes cause us troubles since we don't know exactly which RPMs they will build (so the cache fetcher
+		// will pull all of the possible matches). Since we are already matching against SRPMs, we can safely skip these
+		// nodes since they will be included in another node with the same SRPM.
+		if n.Implicit {
```

Comment:
What scenarios makes us create implicit build nodes? I understood implicit provides but I'm not sure what an implicit build node means.

Reply:
This may be an artifact of an earlier implementation, I'll run some tests and double check if this occurs still. I *think* that implicit nodes will never be build nodes...

Reply:
Ok, so sometimes we get an 'implicit' node (something of the form `foo(bar)` where we DO actually know the source so we can create a build + run node, but other times we have no idea and leave it as an unresolved node. We need to handle both cases here.

Reply:
Actually, thinking about it again we can probably just skip all this logic since the only time we get an implicit build node is when we DO have all the data we need... I'll give it a shot.

Reply:
Removing all special handling for these seems fine.